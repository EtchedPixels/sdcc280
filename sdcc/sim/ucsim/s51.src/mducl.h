/*
 * Simulator of microcontrollers (s51.src/mducl.h)
 *
 * Copyright (C) 2016,16 Drotos Daniel, Talker Bt.
 * 
 * To contact author send email to drdani@mazsola.iit.uni-miskolc.hu
 *
 */

/* This file is part of microcontroller simulator: ucsim.

UCSIM is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

UCSIM is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with UCSIM; see the file COPYING.  If not, write to the Free
Software Foundation, 59 Temple Place - Suite 330, Boston, MA
02111-1307, USA. */
/*@1@*/

#ifndef MDUCL_HEADER
#define MDUCL_HEADER

#include "hwcl.h"


class cl_mdu517: public cl_hw
{
 protected:
  u8_t v[7];
  class cl_memory_cell *regs[7];//, *arcon;
  u64_t writes;
  int nuof_writes;
  //bool calcing;
 public:
  cl_mdu517(class cl_uc *auc, int aid);
  virtual int init(void);
  
  virtual t_mem read(class cl_memory_cell *cell);
  virtual void write(class cl_memory_cell *cell, t_mem *val);
  virtual t_mem conf_op(cl_memory_cell *cell, t_addr addr, t_mem *val);
};


#endif

/* End of s51.src/mducl.h */
