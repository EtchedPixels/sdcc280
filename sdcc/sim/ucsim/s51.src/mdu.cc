/*
 * Simulator of microcontrollers (s51.src/mdu.cc)
 *
 * Copyright (C) 2016,16 Drotos Daniel, Talker Bt.
 * 
 * To contact author send email to drdani@mazsola.iit.uni-miskolc.hu
 *
 */

/* This file is part of microcontroller simulator: ucsim.

UCSIM is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 2 of the License, or
(at your option) any later version.

UCSIM is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with UCSIM; see the file COPYING.  If not, write to the Free
Software Foundation, 59 Temple Place - Suite 330, Boston, MA
02111-1307, USA. */
/*@1@*/

#include "uc51cl.h"

#include "mducl.h"


cl_mdu517::cl_mdu517(class cl_uc *auc, int aid):
  cl_hw(auc, HW_CALC, aid, "mdu")
{
}

int
cl_mdu517::init(void)
{
  int i;
  class cl_51core *u= (cl_51core*)uc;

  cl_hw::init();
  
  //arcon= register_cell(u->sfr, 0xef);
  for (i= 0; i<7; i++)
    {
      regs[i]= register_cell(u->sfr, 0xe9+i);
      v[i]= regs[i]->get();
    }
  nuof_writes= 0;
  writes= 0xffffffffffff;
  //calcing= 0;
  return 0;
}

t_mem
cl_mdu517::read(class cl_memory_cell *cell)
{
  cl_address_space *sfr= ((cl_51core*)uc)->sfr;
  t_addr a;
  t_mem v= cell->get();
  
  if (conf(cell, NULL))
    return v;
  if (sfr->is_owned(cell, &a))
    {
      a-= 0xe9;
      if ((a < 0) ||
	  (a > 6))
	return v;
      //printf("\nREAD a=%ld v=%02x %p\n", a, v, cell);
      if (a == 6)
	cell->set(v & ~0x80);
    }
  return v;
}

void
cl_mdu517::write(class cl_memory_cell *cell, t_mem *val)
{
  cl_address_space *sfr= ((cl_51core*)uc)->sfr;
  t_addr a;
  u8_t ar= regs[6]->get() & ~0x80;
  
  if (conf(cell, val))
    return;
  if (sfr->is_owned(cell, &a))
    {
      // if (a==0xee) printf(" WRITE EE %02x\n", *val);
      a-= 0xe9;
      if ((a < 0) ||
	  (a > 6))
	{
	  return;
	}
      /*if (calcing)
	{
	  regs[6]->set(ar | 0x80);
	  return;
	  }*/
      if (a == 0)
	{
	  writes= 0xffffffffffff;
	  nuof_writes= 0;
	  regs[6]->set(ar & ~0x80);
	}
      if (nuof_writes > 5)
	{
	  regs[6]->set(ar | 0x80);
	  return;
	}
      writes&= ~(0xffL << (nuof_writes*8));
      writes|= ((u64_t)a << (nuof_writes*8));
      v[a]= *val;
      nuof_writes++;
      if (a == 6)
	{
	  writes= 0xff0603020100; // force norm/shift
	  v[a]&= ~0x80;
	  ar= v[6];
	}

      switch (writes)
	{
	  //   665544332211
	case 0x050403020100:
	  {
	    // 32/16
	    u32_t dend= v[3]*256*256*256 + v[2]*256*256 + v[1]*256 + v[0];
	    u16_t dor= v[5]*256 + v[4];
	    u32_t quo= 0;
	    u16_t rem= 0;
	    if (dor == 0)
	      regs[6]->set(ar | 0x40);
	    else
	      {
		quo= dend / dor;
		rem= dend % dor;
		regs[6]->set(ar & ~0x40);
		//printf("\nSIM %u/%u=%u,%u %x,%x\n", dend, dor, quo, rem, quo, rem);
	      }
	    regs[0]->set(quo & 0xff);
	    regs[1]->set((quo>>8) & 0xff);
	    regs[2]->set((quo>>16) & 0xff);
	    regs[3]->set((quo>>24) & 0xff);
	    regs[4]->set(rem & 0xff);
	    regs[5]->set((rem>>8) & 0xff);
	    /*{
	      int j;
	      for (j=0;j<6;j++)
		{
		  printf("  REG[%d]=%02x/%02x %p\n",j,regs[j]->get(),sfr->get(0xe9+j), regs[j]);
		}
		}*/
	    //calcing= 6;
	    writes= 0xffffffffffff;
	    nuof_writes= 0;
	    break;
	  }
	  //   665544332211
	case 0xffff05040100:
	  {
	    // 16/16
	    u16_t dend= v[1]*256 + v[0];
	    u16_t dor= v[5]*256 + v[4];
	    u16_t quo= 0;
	    u16_t rem= 0;
	    if (dor == 0)
	      regs[6]->set(ar | 0x40);
	    else
	      {
		quo= dend / dor;
		rem= dend % dor;
		regs[6]->set(ar & ~0x40);
	      }
	    regs[0]->set(quo & 0xff);
	    regs[1]->set((quo>>8) & 0xff);
	    regs[4]->set(rem & 0xff);
	    regs[5]->set((rem>>8) & 0xff);
	    //calcing= 6;
	    writes= 0xffffffffffff;
	    nuof_writes= 0;
	    break;
	  }
	  //   665544332211
	case 0xffff05010400:
	  {
	    // 16*16
	    u16_t mand= v[1]*256 + v[0];
	    u16_t mor= v[5]*256 + v[4];
	    u32_t pr= mand * mor;
	    regs[0]->set(pr & 0xff);
	    regs[1]->set((pr>>8) & 0xff);
	    regs[2]->set((pr>>16) & 0xff);
	    regs[3]->set((pr>>24) & 0xff);
	    if (pr > 0xffff)
	      regs[6]->set(ar | 0x40);
	    else
	      regs[6]->set(ar & ~0x40);
	    writes= 0xffffffffffff;
	    nuof_writes= 0;
	    break;
	  }
	  //   665544332211
	case 0xff0603020100:
	  {
	    // norm, shift
	    u32_t d;
	    d= v[3]*256*256*256 + v[2]*256*256 + v[1]*256 + v[0];
	    if ((ar & 0x1f) == 0)
	      {
		// normalize
		if (d == 0)
		  regs[6]->set(ar & ~0x1f);
		else if (d & 0x80000000)
		  regs[6]->set(ar | 0x40);
		else
		  {
		    int i;
		    for (i= 0; (d&0x80000000)==0; i++)
		      d<<= 1;
		    regs[6]->set((ar&~0x1f) | i);
		    //printf("NORM d=%x i=%d\n", d, i);
		  }
	      }
	    else
	      {
		// shift
		if (ar & 0x20)
		  d<<= (ar & 0x1f);
		else
		  d>>= (ar & 0x1f);
	      }
	    regs[0]->set(d & 0xff);
	    regs[1]->set((d>>8) & 0xff);
	    regs[2]->set((d>>16) & 0xff);
	    regs[3]->set((d>>24) & 0xff);
	    writes= 0xffffffffffff;
	    nuof_writes= 0;
	    break;
	  }
	default:
	  if (nuof_writes > 5)
	    {
	      regs[6]->set(ar | 0x80);
	      writes= 0xffffffffffff;
	      nuof_writes= 0;
	    }
	  break;
	}
    }
  *val= regs[a]->get();
}

t_mem
cl_mdu517::conf_op(cl_memory_cell *cell, t_addr addr, t_mem *val)
{
  return cell->get();
}


/* End of s51.src/mdu.cc */
