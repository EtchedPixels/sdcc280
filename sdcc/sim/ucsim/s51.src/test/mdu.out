5006 words read from mdu.ihx
32div16     6374/2781 25460/10113=0,2 mdu=2/10098 fail 
32div16 ffff9e59/15fb 4294942297/5627=11,42377 mdu=4294944137/5382 fail 
32div16 ffffd896/3f54 4294957206/16212=4,2780 mdu=2780/16230 fail 
32div16     649e/31df 25758/12767=0,2 mdu=2/12768 fail 
32div16 ffffa590/7de1 4294944144/32225=2,2207 mdu=2207/32209 fail 
32div16 fffff287/52bf 4294963847/21183=3,6147 mdu=6147/21066 fail 
32div16     197e/6956 6526/26966=0,0 mdu=0/27006 fail 
32div16 ffffc3ae/2847 4294951854/10311=6,23324 mdu=23324/10474 fail 
32div16 ffffc40a/4287 4294951946/17031=3,55576 mdu=4294957336/16994 fail 
32div16 ffffde1a/6488 4294958618/25736=2,35813 mdu=4294937573/25714 fail 
32div16     77dc/6eea 30684/28394=0,1 mdu=1/28402 fail 
32div16 fffffe62/4a7b 4294966882/19067=3,28648 mdu=28648/19178 fail 
32div16 ffffcc1a/19a1 4294954010/6561=9,64794 mdu=4294966554/6592 fail 
32div16 ffff86a0/3b54 4294936224/15188=4,20640 mdu=20640/15136 fail 
32div16     529b/1b37 21147/6967=0,3 mdu=3/7158 fail 
32div16     4aa6/345b 19110/13403=0,1 mdu=1/13387 fail 
32div16     61b4/7a79 25012/31353=0,0 mdu=0/31412 fail 
32div16 ffff9d20/60e3 4294941984/24803=2,42090 mdu=4294943850/24610 fail 
32div16 fffff3fc/ 58e 4294964220/1422=46,5712 mdu=5712/1436 fail 
32div16 ffff9287/ 799 4294939271/1945=33,45506 mdu=4294947266/1941 fail 
32div16     3451/493f 13393/18751=0,0 mdu=0/18769 fail 
32div16     5434/2043 21556/8259=0,2 mdu=2/8366 fail 
32div16     2718/3d0d 10008/15629=0,0 mdu=0/15640 fail 
32div16 fffff98c/36a6 4294965644/13990=4,44858 mdu=4294946618/14064 fail 
32div16 ffffb628/ 586 4294948392/1414=46,22789 mdu=22789/1418 fail 
32div16     4608/5929 17928/22825=0,0 mdu=0/22792 fail 
32div16 ffff8036/62ff 4294934582/25343=2,38400 mdu=4294940160/25142 fail 
32div16 fffffc96/1d70 4294966422/7536=8,45638 mdu=4294947398/7670 fail 
32div16     2160/718f 8544/29071=0,0 mdu=0/29024 fail 
32div16 ffffffd8/3305 4294967256/13061=5,1159 mdu=1159/13109 fail 
32div16 ffffecd0/7809 4294962384/30729=2,8697 mdu=8697/30735 fail 
32div16 ffffa254/7bcd 4294943316/31693=2,4445 mdu=4445/31707 fail 
32div16     7e2e/5648 32302/22088=0,1 mdu=1/22246 fail 
32div16     2302/23d3 8962/9171=0,0 mdu=0/8962 ok 
32div16 ffffdb4a/6006 4294957898/24582=2,43647 mdu=4294945407/24656 fail 
32div16 ffffe456/25b5 4294960214/9653=6,51719 mdu=4294953479/9571 fail 
32div16 ffffa08c/4813 4294942860/18451=3,36167 mdu=4294937927/18503 fail 
32div16 fffff658/7eb2 4294964824/32434=2,1349 mdu=1349/32350 fail 
32div16 fffff71e/60ac 4294965022/24748=2,42475 mdu=4294944235/24634 fail 
32div16 ffffce90/7d31 4294954640/32049=2,2940 mdu=2940/32212 fail 
32div16 ffffb9f8/1046 4294949368/4166=15,47912 mdu=4294949672/4104 fail 
32div16     51c5/7e36 20933/32310=0,0 mdu=0/32453 fail 
32div16     6899/1ef6 26777/7926=0,3 mdu=3/7863 fail 
32div16     1575/ a16 5493/2582=0,2 mdu=2/2633 fail 
32div16 fffff2d0/1f74 4294963920/8052=8,9115 mdu=9115/8084 fail 
32div16     7778/7e7c 30584/32380=0,0 mdu=0/32376 fail 
32div16     1e44/5f41 7748/24385=0,0 mdu=0/24388 fail 
32div16     4e9f/4371 20127/17265=0,1 mdu=1/17198 fail 
32div16 ffffe452/2c70 4294960210/11376=5,49865 mdu=4294951625/11362 fail 
32div16 ffffae2f/5b01 4294946351/23297=2,53284 mdu=4294955044/23307 fail 
32div16 fffff9f5/ c33 4294965749/3123=20,64549 mdu=4294966309/3222 fail 
32div16     7b8d/5ec0 31629/24256=0,1 mdu=1/24269 fail 
32div16 fffff1b1/640d 4294963633/25613=2,36614 mdu=4294938374/25699 fail 
32div16 fffffeaf/2433 4294966959/9267=7,4716 mdu=4716/9259 fail 
32div16     7866/485e 30822/18526=0,1 mdu=1/18440 fail 
32div16 ffffc1a2/38e6 4294951330/14566=4,32717 mdu=32717/14452 fail 
32div16 ffffff20/4da6 4294967072/19878=3,19458 mdu=19458/19924 fail 
32div16     6d1a/20c3 27930/8387=0,3 mdu=3/8401 fail 
32div16     14c3/ ea8 5315/3752=0,1 mdu=1/3611 fail 
32div16 ffffa5bc/2cdb 4294944188/11483=5,46346 mdu=4294948106/11310 fail 
32div16     72c0/4860 29376/18528=0,1 mdu=1/18528 fail 
32div16     5d93/4ebf 23955/20159=0,1 mdu=1/20180 fail 
32div16     7f00/2567 32512/9575=0,3 mdu=3/9675 fail 
32div16     10a6/6c25 4262/27685=0,0 mdu=0/27814 fail 
32div16     46a0/4531 18080/17713=0,1 mdu=1/17775 fail 
32div16 ffffae53/5862 4294946387/22626=2,58751 mdu=4294960511/22709 fail 
32div16     2a48/684f 10824/26703=0,0 mdu=0/26696 fail 
32div16 fffffaa4/2d09 4294965924/11529=5,44855 mdu=4294946615/11701 fail 
32div16     34dc/219f 13532/8607=0,1 mdu=1/8509 fail 
32div16 ffff98f0/30a9 4294940912/12457=5,17101 mdu=17101/12443 fail 
32div16     378c/2d0b 14220/11531=0,1 mdu=1/11649 fail 
32div16 ffff88a0/4280 4294936736/17024=3,55679 mdu=4294957439/16928 fail 
32div16 fffff42a/4747 4294964266/18247=3,38771 mdu=4294940531/18245 fail 
32div16     57c8/4ba5 22472/19365=0,1 mdu=1/19235 fail 
32div16 ffffc570/46ac 4294952304/18092=3,40787 mdu=4294942547/18092 fail 
32div16 ffffe7e0/71eb 4294961120/29163=2,16202 mdu=16202/29170 fail 
32div16      87f/7424 2175/29732=0,0 mdu=0/29823 fail 
32div16 ffff9f96/2692 4294942614/9874=6,41758 mdu=4294943518/9850 fail 
32div16 ffffc400/6bc5 4294951936/27589=2,24604 mdu=24604/27508 fail 
32div16     604a/1a85 24650/6789=0,3 mdu=3/6843 fail 
32div16     72b1/4f8c 29361/20364=0,1 mdu=1/20261 fail 
32div16     1840/13ab 6208/5035=0,1 mdu=1/5013 fail 
32div16 ffff8a4c/ 352 4294937164/850=77,6595 mdu=6595/982 fail 
32div16 fffffe72/ 29c 4294966898/668=98,7063 mdu=7063/622 fail 
32div16 ffffe65f/2ff9 4294960735/12281=5,22044 mdu=22044/12067 fail 
32div16     5c6c/7dba 23660/32186=0,0 mdu=0/32108 fail 
32div16 fffff520/1c5d 4294964512/7261=9,1687 mdu=1687/7237 fail 
32div16 ffffac1b/5e7e 4294945819/24190=2,46478 mdu=4294948238/24119 fail 
32div16 ffffea6e/358c 4294961774/13708=4,51173 mdu=4294952933/13618 fail 
32div16     5c7f/1771 23679/6001=0,3 mdu=3/5932 fail 
32div16     5dde/145e 24030/5214=0,4 mdu=4/5222 fail 
32div16      692/6533 1682/25907=0,0 mdu=0/26002 fail 
32div16 ffffff60/1d51 4294967136/7505=8,47992 mdu=4294949752/7528 fail 
32div16 ffffcb26/2a7b 4294953766/10875=6,1722 mdu=1722/10952 fail 
32div16     6113/3064 24851/12388=0,2 mdu=2/12363 fail 
32div16 ffff8a2a/ 6ef 4294937130/1775=36,60386 mdu=4294962146/1580 fail 
32div16 ffffa04a/7142 4294942794/28994=2,17060 mdu=17060/28930 fail 
32div16 ffffe420/7a51 4294960160/31313=2,6090 mdu=6090/31286 fail 
32div16     7a8a/2cf7 31370/11511=0,2 mdu=2/11420 fail 
32div16     2da6/1945 11686/6469=0,1 mdu=1/6497 fail 
32div16 test: succ=1 fails=99
